### PR TITLE
jep: new recipe

### PR DESCRIPTION
**Pull request by tschoonj**
@conda-forge/help-python-c @conda-forge/help-java

Going to need your help on this one: I need a JDK to build this thing.

Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by tschoonj** [comment]
Any thoughts on how I get the java SDK into the Linux build? Yum?

**Comment by tschoonj** [comment]
Any thoughts on how to fix these errors on Linux?

```
Warning: rpath /home/conda/staged-recipes/build_artifacts/jep_1573227428002/_build_env/lib is outside prefix /home/conda/staged-recipes/build_artifacts/jep_1573227428002/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho (removing it)
Warning: rpath /home/conda/staged-recipes/build_artifacts/jep_1573227428002/_build_env/lib is outside prefix /home/conda/staged-recipes/build_artifacts/jep_1573227428002/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho (removing it)
  ERROR :: $RPATH/libjvm.so not in prefix_owners
  ERROR :: $RPATH/libjvm.so not in prefix_owners
   INFO (jep,lib/python2.7/site-packages/jep/libjep.so): Needed DSO x86_64-conda_cos6-linux-gnu/sysroot/lib/libdl.so.2 found in CDT/compiler package conda-forge::gcc_impl_linux-64-7.3.0-habb00fd_2
   INFO (jep,lib/python2.7/site-packages/jep/libjep.so): Needed DSO lib/libpython2.7.so.1.0 found in conda-forge::python-2.7.15-h5a48372_1009
  ERROR (jep,lib/python2.7/site-packages/jep/libjep.so): $RPATH/libjvm.so not found in sysroot, is this binary repackaging? .. do you need to use install_name_tool/patchelf?
   INFO (jep,lib/python2.7/site-packages/jep/libjep.so): Needed DSO x86_64-conda_cos6-linux-gnu/sysroot/lib/libpthread.so.0 found in CDT/compiler package conda-forge::gcc_impl_linux-64-7.3.0-habb00fd_2
   INFO (jep,lib/python2.7/site-packages/jep/libjep.so): Needed DSO x86_64-conda_cos6-linux-gnu/sysroot/lib/libc.so.6 found in CDT/compiler package conda-forge::gcc_impl_linux-64-7.3.0-habb00fd_2
   INFO (jep,lib/python2.7/site-packages/jep/jep.so): Needed DSO x86_64-conda_cos6-linux-gnu/sysroot/lib/libdl.so.2 found in CDT/compiler package conda-forge::gcc_impl_linux-64-7.3.0-habb00fd_2
   INFO (jep,lib/python2.7/site-packages/jep/jep.so): Needed DSO lib/libpython2.7.so.1.0 found in conda-forge::python-2.7.15-h5a48372_1009
  ERROR (jep,lib/python2.7/site-packages/jep/jep.so): $RPATH/libjvm.so not found in sysroot, is this binary repackaging? .. do you need to use install_name_tool/patchelf?
   INFO (jep,lib/python2.7/site-packages/jep/jep.so): Needed DSO x86_64-conda_cos6-linux-gnu/sysroot/lib/libpthread.so.0 found in CDT/compiler package conda-forge::gcc_impl_linux-64-7.3.0-habb00fd_2
   INFO (jep,lib/python2.7/site-packages/jep/jep.so): Needed DSO x86_64-conda_cos6-linux-gnu/sysroot/lib/libc.so.6 found in CDT/compiler package conda-forge::gcc_impl_linux-64-7.3.0-habb00fd_2
WARNING (jep): dso library package conda-forge::openjdk-8.0.192-h14c3975_1003 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
Traceback (most recent call last):
```

**Comment by jan-janssen** [comment]
At the moment you are just copying files, would not it make more sense to compile Firefox from source?

**Comment by tschoonj** [comment]
@conda-forge/help-python-c 
Could someone help me out here? Thanks!

**Comment by beckermr** [comment]
Sorry for the delay. I don’t personally understand them. Is that library made by this package? You may need to adjust the linking information by hand or this might be an issue with some upstream dep.

**Comment by tschoonj** [comment]
The `libjvm.so` library that is referred to in the error is part of `openjdk`, not `jep`... not sure why it is complaining about sysroot...

**Comment by beckermr** [comment]
Ok. So the complaint appears to be that the location at which jep is linking to that file is outside of the conda prefix. This is due to something in the build of the package.

**Comment by tschoonj** [comment]
Could this be because this library is installed in `$CONDA_PREFIX/jre/lib/amd64/server/libjvm.so`?

**Comment by beckermr** [comment]
Well the link location should still reflect that path, not what is in the error

File: recipes/jep/fix-split.patch [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+index ccbda7e..f6b600f 100644
+--- a/commands/java.py
++++ b/commands/java.py
+@@ -208,7 +208,9 @@ class setup_java(Command):
+                     'INFO: the MACOSX_DEPLOYMENT_TARGET environment variable is set:', target)
+ 
+                 result = shell('sw_vers -productVersion')
+-                if target.split('.')[:2] != result.stdout.split('.')[:2]:
++                print("type target: {}".format(type(target)))
++                print("type result.stdout: {}".format(type(result.stdout)))
```
Comment (beckermr) [inline recipes/jep/fix-split.patch]: Do you want these print statements in the patches?

File: recipes/jep/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  script: "{{ PYTHON }} -m pip install . -vv"
+  ignore_run_exports:
+    - openjdk
+    - numpy
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+  host:
+    - openjdk=8
```
Comment (beckermr) [inline recipes/jep/meta.yaml]: You need a space here.

File: recipes/jep/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    - {{ pin_compatible('numpy') }}
+
+about:
+  home: "https://github.com/ninia/jep"
+  license: "zlib/libpng"
+  license_family: OTHER
+  license_file: LICENSE
+  summary: "Jep embeds CPython in Java"
+  doc_url: 
+  dev_url:
```
Comment (beckermr) [inline recipes/jep/meta.yaml]: Remove these empty keys.

File: recipes/jep/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    - python
+    - numpy
+  run:
+    - openjdk
+    - python
+    - {{ pin_compatible('numpy') }}
+
+about:
+  home: "https://github.com/ninia/jep"
+  license: "zlib/libpng"
```
Comment (beckermr) [inline recipes/jep/meta.yaml]: Is this the license https://spdx.org/licenses/zlib-acknowledgement.html?

If yes, then use this name for it

```
zlib-acknowledgement
```

**Comment by tschoonj** [comment]
Think I implemented all your requested changes... now I need to write a test for it!

File: recipes/jep/fix-split.patch [hunk 5]
```diff
@@ -13,8 +13,6 @@ index ccbda7e..f6b600f 100644
  
                  result = shell('sw_vers -productVersion')
 -                if target.split('.')[:2] != result.stdout.split('.')[:2]:
-+                print("type target: {}".format(type(target)))
-+                print("type result.stdout: {}".format(type(result.stdout)))
```
Comment (beckermr) [inline recipes/jep/fix-split.patch]: This may mess up the patch since the number of lines changed. I am not 100% sure.

Comment (beckermr) [reply recipes/jep/fix-split.patch]: Looks like it did. Glad you fixed it!

**Comment by tschoonj** [comment]
Stuck again when running `jep` on Linux:

```bash
$ jep
ERROR: ld.so: object '/dls_sw/apps/python/anaconda/4.6.14/64/envs/openjdk-test/lib/libpython3.8.a' from LD_PRELOAD cannot be preloaded: ignored.

              No readline available. History will not be available.
              
                   You may want to set the LD_PRELOAD environment variable, see the
                   README file for details.

                   i.e.: export LD_PRELOAD=/usr/lib/libpython2.7.so.1.0
                   
>>>
```
This can be traced back to:

https://github.com/ninia/jep/blob/3bcf611b3bc77328cbb38d4f49eec1280602175e/commands/python.py#L37-L44

The conda pythons for Linux have `LDLIBRARY` set through an override to the [python static archive]( 'LDLIBRARY': 'libpython@PYVER@@ABIFLAGS@.a',) instead of the shared library. This is not cool, as has been pointed out [before](https://github.com/conda-forge/python-feedstock/issues/181). I don't even understand why: Ubuntu also statically links the python interpreters against the library, and `LDLIBRARY` and `INSTSONAME` return the shared library name.

**Comment by beckermr** [comment]
@isuruf

**Comment by tschoonj** [comment]
@isuruf could you comment on this one please?

Thanks!

**Comment by isuruf** [comment]
reopened https://github.com/conda-forge/python-feedstock/issues/181. This should be fixed there

**Pull request closed without merging**